### ISSUE TITLE
Since WINE-8.x, loading ReShade makes WINE crash with "Assertion failed: !csw->is32" error

### ISSUE THREAD

**Issue by gabriele2000** [doitsujin/dxvk#3677]
IMPORTANT: If it happens to you and you're reading this, scroll down or just [click here](https://github.com/doitsujin/dxvk/issues/3677#issuecomment-1763183439) to go to the relevant comment!

UPDATE: DXVK is to blame https://github.com/doitsujin/dxvk/issues/3677#issuecomment-1742049453 also the [relevant PR](https://github.com/doitsujin/dxvk/pull/3635)

EDIT: I've edited the issue to focus on ReShade, the issue with the addon will be covered in a future issue

Loading a game with ReShade's dll too causes WINE to crash with a weird error:
`Assertion failed: !csw->is32, file /home/builder/build_wine/wine/dlls/dbghelp/cpu_x86_64.c, line 650`

### System information
- GPU: Nvidia GTX 1050TI
- Driver: 535.113.01
- Wine version: 7.20 works with DXVK-2.3, 8.x works only with DXVK-2.2
- DXVK version: 2.3
- Pop!_Os 22.04

### Log files
- [Proton.log](https://github.com/doitsujin/dxvk/files/12764023/steam-heroic-0.log)

**Comment by HunterCZ122** [comment]
Just to make sure, did you run winetricks d3dcompiler_47 (or get d3dcompiler_47.dll from other game and put it next to bf4.exe) on your wine prefix? ReShade iirc requires native d3dcompiler_47.dll to work.

**Comment by gabriele2000** [comment]
> Just to make sure, did you run winetricks d3dcompiler_47 (or get d3dcompiler_47.dll from other game and put it next to bf4.exe) on your wine prefix? ReShade iirc requires native d3dcompiler_47.dll to work.

Yes, I have native d3dcompiler_47 in system32 and syswow64, otherwise with wine 7.20 it wouldn't work.

There's a little problem by the way, I've been having this weird problem since half a month.
Sometimes wine crashes with the error `Assertion failed: !csw->is32, file /home/builder/build_wine/wine/dlls/dbghelp/cpu_x86_64.c, line 650`

**Comment by Stereo-3D** [comment]
I also have similar issues with some anime game but let me check if our situation is really similar:
- Is removing `dxvk-cache` file before launching the game fix the problem temporarily?
- When using older dxvk version release (for example: DXVK 2.2), is it has the exact same error/problem?

**Comment by gabriele2000** [comment]
>     * Is removing `dxvk-cache` file before launching the game fix the problem temporarily?

YES, It did fix the problem once, I've also tried with deleting nvidia shader cache, I tried once and it worked.

>     * When using older dxvk version release (for example: DXVK 2.2), is it has the exact same error/problem?

I don't remember if I've tried, but I will try using DXVK 2.2 and report back in less than one hour

**Comment by gabriele2000** [comment]
Bisecting the problem leads to this:

<pre>428ca9416d60cf7025ae97da88f44eabc2ec0aa2 is the first bad commit
<font color="#C4A000">commit 428ca9416d60cf7025ae97da88f44eabc2ec0aa2</font>
Author: Joshua Ashton &lt;joshua@froggi.es&gt;
Date:   Wed Aug 23 14:50:42 2023 +0100

    [d3d11] Implement synchronization on keyed mutexes
    
    Co-authored-by: Paul Gofman &lt;pgofman@codeweavers.com&gt;

 src/d3d11/d3d11_context_imm.h   |  1 <font color="#4E9A06">+</font>
 src/d3d11/d3d11_initializer.cpp | 12 <font color="#4E9A06">++++++++++++</font>
 src/d3d11/d3d11_initializer.h   |  2 <font color="#4E9A06">++</font>
 src/d3d11/d3d11_resource.cpp    | 36 <font color="#4E9A06">+++++++++++++++++++++++++++++++</font><font color="#CC0000">-----</font>
 src/d3d11/d3d11_resource.h      |  2 <font color="#4E9A06">++</font>
 5 files changed, 48 insertions(+), 5 deletions(-)
</pre>

**Comment by Stereo-3D** [comment]
Wow good work, I never investigate this far into the issue, it seems that we have the exact same issue (although on different game).. I just switched back to DXVK 2.2, continue using ReShade with REST plugin and pretend this never happen, hehe.. Thanks for reporting this, I wish the issue will be resolved soon :pray:

**Comment by gabriele2000** [comment]
> Wow good work, I never investigate this far into the issue, it seems that we have the exact same issue (although on different game).. I just switched back to DXVK 2.2, continue using ReShade with REST plugin and pretend this never happen, hehe.. Thanks for reporting this, I wish the issue will be resolved soon 🙏

Latest REST versions don't work though... plus, going back to DXVK-2.2 is simply not so acceptable, the drawbacks compared to v2.3 are simply too many

**Comment by gabriele2000** [comment]
> UPDATE: DXVK is to blame [#3677 (comment)](https://github.com/doitsujin/dxvk/issues/3677#issuecomment-1742049453) also [relevant PR](https://github.com/doitsujin/dxvk/pull/3635/)
> 
> EDIT: I've edited the issue to focus on ReShade, the issue with the addon will be covered in a future issue
> 
> Loading a game with ReShade's dll too causes WINE to crash with a weird error: `Assertion failed: !csw->is32, file /home/builder/build_wine/wine/dlls/dbghelp/cpu_x86_64.c, line 650`
> ### System information
> 
>     * GPU: Nvidia GTX 1050TI
> 
>     * Driver: 535.113.01
> 
>     * Wine version: 7.20 works with DXVK-2.3, 8.x works only with DXVK-2.2
> 
>     * DXVK version: 2.3
> 
>     * Pop!_Os 22.04
> 
> 
> ### Log files
> 
>     * [Proton.log](https://github.com/doitsujin/dxvk/files/12764023/steam-heroic-0.log)

**Comment by Blisto91** [comment]
Can you retest with latest master? A reshade crash regression was indeed found and fixed since 2.3.
You can download a pre compiled build here https://github.com/doitsujin/dxvk/actions/runs/6240574600

**Comment by gabriele2000** [comment]
> Can you retest with latest master? A reshade crash regression was indeed found and fixed since 2.3. You can download a pre compiled build here https://github.com/doitsujin/dxvk/actions/runs/6240574600

That is a nice surprise! It works... for now.
Let's see if it didn't break other stuff, testing now, will update the comment.

Also @Stereo-3D (sorry for another tag) try this!

**Comment by gabriele2000** [comment]
> Can you retest with latest master? A reshade crash regression was indeed found and fixed since 2.3. You can download a pre compiled build here https://github.com/doitsujin/dxvk/actions/runs/6240574600

With DXVK-2.3 I also had this weird error with Call of Duty: Black Ops 3
`Assertion failed: !status && "vkDestroyBufferView", file ../dlls/winevulkan/loader_thunks.c, line 3226`

something that with DXVK-2.2 wasn't happening, I'll try to bisect that too.

**Comment by Stereo-3D** [comment]
> Can you retest with latest master? A reshade crash regression was indeed found and fixed since 2.3. You can download a pre compiled build here https://github.com/doitsujin/dxvk/actions/runs/6240574600

Yes, I can confirm that this pre compiled build is working with ReShade 5.9.2 (+REST addon) loaded on anime game I'm playing! Thank you very much @Blisto91 and @gabriele2000 :pray:

**Comment by gabriele2000** [comment]
> Yes, I can confirm that this pre compiled build is working with ReShade 5.9.2 (+REST addon) loaded on anime game I'm playing! Thank you very much @Blisto91 and @gabriele2000 🙏

Hell yeah!

**Comment by Blisto91** [comment]
Good to hear.

@gabriele2000 Please open news issues for anything else you think is a dxvk bug.

**Comment by gabriele2000** [comment]
> Good to hear.
> 
> @gabriele2000 Please open news issues for anything else you think is a dxvk bug.

Of course, I will reference the relevant comment in a new issue once I succesfully bisect the relevant commit... plus I have to be certain that it's a DXVK bug since sometimes it's a random error.

**Comment by Blisto91** [comment]
@gabriele2000 Was all good for this particular issue?

**Comment by gabriele2000** [comment]
> @gabriele2000 Was all good for this particular issue?

Upon further testing I'm not seeing any problem since when I claimed that the problem was fixed.

**IMPORTANT**
I want to write something that will be useful to everyone.

Since Reshade 5.9 something tells me that the particular error message `Assertion failed: !csw->is32, file /home/builder/build_wine/wine/dlls/dbghelp/cpu_x86_64.c, line 650` is spitted thanks to a quirk of ReShade.

I've reason to believe this, since (it happens even with the commit that fixes the issue) if a game can't load ReShade properly without using a native d3d_compiler47 dll.
As you might know, ReShade needed the native dll (in the past, before WINE-8.10 or something, now I guess it doesn't anymore but there are games that don't accept that, causing the issue.

Basically ReShade (or the hooking process for that particular game?) does something "illegal" and triggers a check in WINE, introduced years ago.
**If it happens**, place a **native d3d_compiler47.dll** next to the game's executable then:
- `winecfg`
- In the `library` tab, add a custom rule for `d3d_compiler47.dll` as `native, builtin`
- Click on "OK" or "Apply" then "OK"
